File: Cocoa/Classes/MacRomanFormatter.m
#import "MacRomanFormatter.h"
#import <AppKit/AppKit.h>

@implementation MacRomanFormatter
@synthesize stringLength;
@synthesize valueRequired;
@synthesize exactLengthRequired;

- (NSString *)stringForObjectValue:(id)object {
    return object;
}

- (BOOL)getObjectValue:(id *)object forString:(NSString *)string errorDescription:(NSString **)error {
    if (valueRequired && string.length == 0) {
        if (error) {
            *error = @"The value must be not be blank.";
        }
        return NO;
    }
    if (exactLengthRequired && string.length != stringLength) {
        if (error) {
            *error = [NSString stringWithFormat:@"The value must be exactly %d characters.", stringLength];
        }
        return NO;
    }
    if (![string canBeConvertedToEncoding:NSMacOSRomanStringEncoding]) {
        if (error)
            *error = @"The value contains invalid characters for Mac OS Roman encoding.";
        return NO;
    }
    *object = string;
    return YES;
}

- (BOOL)isPartialStringValid:(NSString **)partialStringPtr
       proposedSelectedRange:(NSRangePointer)proposedSelRangePtr
              originalString:(NSString *)origString
       originalSelectedRange:(NSRange)origSelRange
            errorDescription:(NSString **)error
{
    if ([*partialStringPtr length] > stringLength) {
        // If a range is selected then characters in that range will be removed so adjust the insert length accordingly
        NSInteger insertLength = stringLength - origString.length + origSelRange.length;
        
        // Assemble the string
        NSString *prefix = [origString substringToIndex:origSelRange.location];
        NSString *insert = [*partialStringPtr substringWithRange:NSMakeRange(origSelRange.location, insertLength)];
        NSString *suffix = [origString substringFromIndex:origSelRange.location + origSelRange.length];
        *partialStringPtr = [[prefix stringByAppendingString:insert] stringByAppendingString:suffix];
        
        // Fix-up the proposed selection range
        proposedSelRangePtr->location = origSelRange.location + insertLength;
        proposedSelRangePtr->length = 0;
        NSBeep();
        return NO;
    }
    
    return YES;
}

@end

File: RFSupport/MacRomanFormatter.swift
import AppKit

public class MacRomanFormatter: Formatter {
    @IBInspectable public var stringLength: Int = 0
    @IBInspectable public var valueRequired: Bool = false
    @IBInspectable public var exactLengthRequired: Bool = false
    
    public convenience init(stringLength: Int = 0, valueRequired: Bool = false, exactLengthRequired: Bool = false) {
        self.init()
        self.stringLength = stringLength
        self.valueRequired = valueRequired
        self.exactLengthRequired = exactLengthRequired
    }
    
    public override func string(for obj: Any?) -> String? {
        return obj as? String
    }
    
    public override func getObjectValue(_ obj: AutoreleasingUnsafeMutablePointer<AnyObject?>?,
                                        for string: String,
                                        errorDescription error: AutoreleasingUnsafeMutablePointer<NSString?>?) -> Bool {
        if string.isEmpty {
            if valueRequired {
                error?.pointee = NSLocalizedString("The value must be not be blank.", comment: "") as NSString
                return false
            }
        } else if exactLengthRequired && string.count != stringLength {
            error?.pointee = String(format: NSLocalizedString("The value must be exactly %d characters.", comment: ""), stringLength) as NSString
            return false
        }
        if !string.canBeConverted(to: .macOSRoman) {
            error?.pointee = NSLocalizedString("The value contains invalid characters for Mac OS Roman encoding.", comment: "") as NSString
            return false
        }
        obj?.pointee = string as AnyObject
        return true
    }
    
    public override func isPartialStringValid(_ partialStringPtr: AutoreleasingUnsafeMutablePointer<NSString>,
                                              proposedSelectedRange proposedSelRangePtr: NSRangePointer?,
                                              originalString origString: String,
                                              originalSelectedRange origSelRange: NSRange,
                                              errorDescription error: AutoreleasingUnsafeMutablePointer<NSString?>?) -> Bool {
        if partialStringPtr.pointee.length > stringLength {
            // If a range is selected then characters in that range will be removed so adjust the length accordingly
            var range = origSelRange
            range.length += max(stringLength - origString.count, 0)

            // Perform the replacement
            let insert = partialStringPtr.pointee.substring(with: range)
            partialStringPtr.pointee = (origString as NSString).replacingCharacters(in: origSelRange, with: insert) as NSString
            
            // Fix-up the proposed selection range
            proposedSelRangePtr?.pointee.location = range.location + range.length
            proposedSelRangePtr?.pointee.length = 0
            NSSound.beep()
            return false
        }
        
        return true
    }
}
